templates/start-dev-env.sh

```diff
@@ -0,0 +1,133 @@
+DEV_SECRET_JSON=$(kubectl get secret ${DEV_SECRET_NAME} -n ${NAMESPACE} -o json)
+[[ -z "${DEV_SECRET_JSON}" ]] && error_exit "The secret ${DEV_SECRET_NAME} is not existing. Please check secrets with 'kubectl get secrets'"
+
+# Check installations
+if ! command_exist skaffold || ! command_exist telepresence; then
+    if ! command_exist skaffold; then
+        error_exit "command 'skaffold' not found: please visit https://skaffold.dev/docs/install/"
+    fi
+    if ! command_exist kubectl; then
+        error_exit "command 'telepresence' not found. You can download it at https://www.telepresence.io/reference/install."
+    fi
+fi
+
+# Setup dev namepsace
+DEV_USER_PROJECT=${USER_NAME}-${DEV_PROJECT_ID}
+DEV_NAMESPACE=${DEV_USER_PROJECT}
+kubectl get namespace ${DEV_NAMESPACE} >& /dev/null || \
+    (can_i "create namespace" && kubectl create namespace ${DEV_NAMESPACE} && \
+    can_i "create deployment -n ${DEV_NAMESPACE}" && \
+    can_i "create ingress -n ${DEV_NAMESPACE}" && \
+    can_i "create service -n ${DEV_NAMESPACE}" && \
+    can_i "create secret -n ${DEV_NAMESPACE}" && \
+    can_i "create configmap -n ${DEV_NAMESPACE}")
+echo "  Namespace: ${DEV_NAMESPACE}"
+
+# Setup dev secret from pre-configed one
+kubectl get secret ${SECRET_NAME} -n ${DEV_NAMESPACE} >& /dev/null || \
+    echo ${DEV_SECRET_JSON} | jq 'del(.metadata["namespace","creationTimestamp","resourceVersion","selfLink","uid"])' | sed "s/${DEV_SECRET_NAME}/${SECRET_NAME}/g" | kubectl apply -n ${DEV_NAMESPACE} -f -
+echo "  Secret: ${SECRET_NAME}"
+
+# Setup dev service account from pre-configured one
+SERVICE_ACCOUNT=backend-service
+kubectl get sa ${SERVICE_ACCOUNT} -n ${DEV_NAMESPACE} >& /dev/null || \
+    kubectl get sa ${SERVICE_ACCOUNT} -n ${NAMESPACE} -o json | jq 'del(.metadata["namespace","creationTimestamp","resourceVersion","selfLink","uid"])' | kubectl apply -n ${DEV_NAMESPACE} -f -
+
+# Setup dev k8s manifests, configuration, docker login etc
+CONFIG_ENVIRONMENT="staging"
+DEPLOYMENT=${PROJECT_NAME}
+REPO=${PROJECT_NAME}
+EXT_HOSTNAME=<% index .Params `stagingBackendSubdomain`  %><% index .Params `stagingHostRoot` %>
+MY_EXT_HOSTNAME=${DEV_USER_PROJECT}-${EXT_HOSTNAME}
+ECR_REPO=${ACCOUNT_ID}.dkr.ecr.${REGION}.amazonaws.com/${REPO}
+VERSION_TAG=lastest-${USER_NAME}
+DATABASE_NAME=<% index .Params `databaseName` %>
+DEV_DATABASE_NAME=$(echo "dev-${USER_NAME}" | tr -dc 'A-Za-z0-9')
+rm -rf kubernetes.tmp && cp -pr kubernetes kubernetes.tmp
+## image: set image definition
+sed -i ".bak" "s|image: fake-image|image: ${ECR_REPO}:${VERSION_TAG}|g" kubernetes.tmp/base/deployment.yml
+## ingress: create new domain with external DNS
+sed -i ".bak" "s|${EXT_HOSTNAME}|${MY_EXT_HOSTNAME}|g" kubernetes.tmp/overlays/${CONFIG_ENVIRONMENT}/ingress.yml
+## database_name: replace with new one
+sed -i ".bak" "s|DATABASE_NAME=${DATABASE_NAME}|DATABASE_NAME=${DEV_DATABASE_NAME}|g" kubernetes.tmp/base/kustomization.yml
+skaffold config set default-repo ${ECR_REPO} >& /dev/null
```

Comment:
There must be a better way to do this, maybe dynamically within skaffold itself.. We shouldn't be leaving this temp dir hanging around.

Reply:
decided to use kustomize | sed instead of skaffold

Reply:
(from Bill) We also need to think about how the frontend works with this. I guess the idea is that they run the frontend locally but then they point it at myname-app.backend.com. But in that case we also need to make sure the backend has CORS set up properly to allow traffic from the local frontend.

Reply:
when they run the frontend locally, suggest they set /etc/host with 127.0.01 to `app.backend.com` and point api server with `xxx-api.backend.com`. Then, we don't need to change CORS.

Reply:
No, we don't want them to mess with their /etc/hosts, plus that would mess up any use of certificates. It's easy enough for us to set cors headers through the ingress definition.

Reply:
Decided to add `localhost` to CORS. However, annonatation `cors-allow-origin` doesn't support multiple domains in the field, ref: https://github.com/kubernetes/ingress-nginx/issues/5496